### ISSUE TITLE
Bug Report: Flatten Behaviour

### ISSUE THREAD

**Issue by perrelet** [mfeemster/fractorium#85]
Hey there,

Thanks again for the amazing work - just dropping a bug report in the bug trap.

Some strange behavior with flattening xForm 1 on flames upon load. I am aware of the flatten / unflatten options, however these options are not workable for multiple flames simultaneously. Testing, I have found that when saving a set of flames that have been incorrectly flattened the saved xml will be unflattened (correct).

I have found this error when loading flames sculpted in JWF, haven't tested other apps yet.

Hope it helps,

Happy Fractalising

**Comment by mfeemster** [comment]
Please provide example flames with repro steps.

**Comment by perrelet** [comment]
Here's an example flame (created in JWF).
It shouldn't be flat, but loads as flat.
Not a big issue, but can be challenging when loading these as a batch...

[flame.zip](https://github.com/mfeemster/fractorium/files/177505/flame.zip)

**Comment by mfeemster** [comment]
I've spotted a weird bug which your flame revealed:

1) If you load it, then flatten, then unflatten, you don't get the original back. This is because the flatten variation in the first xform has a weight of 0. I am going to add a fix to only add variations with non-zero weights because the entire codebase relies on the presence of a variation meaning that its weight is non-zero. Thanks for finding that.

2) Regarding batch operations. Would you like me to add a menu items that perform "flatten all" and "unflatten all" which will operate on all currently opened flames?

3) Here is arbitrary flatten logic copied Apo which I mentioned earlier:

If an xform does not have these variations in it, then add the flatten variation:

                                "pre_crop",
				"pre_falloff2",
				"pre_rotate_x",
				"pre_rotate_y",
				"pre_ztranslate",
				"blur3D",
				"bubble",
				"bwraps",
				"bwraps2",
				"crop",
				"cylinder",
				"falloff2",
				"hemisphere",
				"julia3D",
				"julia3Dz",
				"linear3D",
				"zblur",
				"zcone",
				"ztranslate",
				"post_crop",
				"post_falloff2",
				"post_rotate_x",
				"post_rotate_y",
				"curl3D_cz",

It all seems rather arbitrary, but that's what it is.

4) I only apply the auto-flattening operation on load if it came from a program other than Fractorium as specified in the version Xml field.

5) I am putting this all here to remind myself to document it on the site.

**Comment by perrelet** [comment]
Great stuff.

1) Indeed, just don't add variations that have for zero weights.

2) Unflatten All / Flatten All would be useful. Although perhaps down the line you might consider other bulk operations also. For example: Set brightness for all flames. I say this because I have had issues with loading bulk flames and their apparent brightness being over exposed (compared to the same .flame loaded in other apps). However, thinking about it now, this issue may very well actually be caused by what is detailed in (6) below.

3) Bizarrely arbitrary...?
4) Good to know.
5) Great.

6) JWF writes the xml attribute "zoom" as "cam_zoom". Apo reads either "zoom" or "cam_zoom" and writes as "zoom". For compatibility sake, might you consider reading the "cam_zoom" attribute as "zoom"?

J :)

**Comment by perrelet** [comment]
This is good stuff.

"Flatten All" / "Unflatten All" would be useful, although it would be amazingly useful if it were possible to apply other parameter changes to the entire set.. For example change the dimensions for all flames in the set.

Hope dev is going well!

**Comment by mfeemster** [comment]
That is doable, I think I can squeeze it in if it doesn't take me too much time.

The only catch is that I won't automatically update all thumbnails in the library tree on every change if applied to all. Instead, the user will just need to click render previews manually.

**Comment by mfeemster** [comment]
In an earlier post, you mentioned being able to parse "cam_zoom" for "zoom", but then you deleted it. I've implemented that with the latest commit.

**Comment by perrelet** [comment]
That would be amazing, but for sure understand that it might be time consuming to write.

Re: the other post - indeed it has disappeared? Not sure why, but glad you saw it any way! Thanks for sorting the cam_zoom thing, that'd be really helpful...

Will be downloading the source in the coming time and having a tinker, so might be able to help take the load off you... Blessings!

**Comment by mfeemster** [comment]
I got to thinking about this and it wouldn't really make sense to apply all to every single field in the program. For example, adding a new variation to the 5th xform in the current flame... how would you apply that to other flames which only have 3 xforms?

So I think it makes most sense to allow for the fields in the Flame and Palette tabs to be applied to all. I could place a checkbox on the top of each one to make it clear.

What do you think?

**Comment by perrelet** [comment]
Absolutely, had the same thoughts - it certainly would not be appropriate in many cases. Also I have concerns with it overcrowding the GUI. What about a simple right click feature "apply to all" instead of a check box?

**Comment by perrelet** [comment]
In testing I've unveiled a few other bits and pieces. Would you like me to post them hear or start a fresh thread for each?

**Comment by mfeemster** [comment]
This won't affect the GPU because the thumbnails are done one the CPU.

If you have other issues, please post them in separate threads. It gets impossible to maintain if they're all in one.

Thanks.

**Comment by perrelet** [comment]
Hey, think you misread that one :) I wrote GUI not GPU...

**Comment by mfeemster** [comment]
Haha, good catch. Sorry about that, I was reading it early in the morning.

You are right to be concerned that updating all thumbnails on every edit could slow things down, especially with a large file, such as one used to make an animation.

I would mitigate that effect by not auto updating the thumbnails. The user would have to manually start the refresh of the thumbnails under the Tools menu. If this were a common occurrence, I'd add it to the toolbar as well.

**Comment by mfeemster** [comment]
I've just done a commit that allows for any adjustment in the Flame tab to be applied to all flames in the current file.

I didn't apply this to "Flatten all". Do you think I should apply it there as well?

**Comment by mfeemster** [comment]
Never mind, I just went ahead and added it. Are you building from source? If so, can you build and let me know if this does what you want so I can close this item? Thanks.

**Comment by perrelet** [comment]
Hey there,

Thanks for all the changes! I was trying to build from source a couple of weeks ago but hit some snags with one of the dependencies. Think I got it sorted - will give it another go at building it...

:)

**Comment by mfeemster** [comment]
Everything should build right out of the box if you follow the build instructions exactly. If it's not building, something is wrong so please report back.
